File: leetcode/python/215-kth-largest-element-in-an-array.py
# Author: RT
# Date: 2022-06-22T14:21:54.515Z
# URL: https://leetcode.com/problems/kth-largest-element-in-an-array/


class Solution:
    def findKthLargest(self, nums: list[int], k: int) -> int:
        n = len(nums)

        def quickselect(l, r):
            ptr = l
            for i in range(l, r):
                if nums[i] <= nums[r]:
                    nums[ptr], nums[i] = nums[i], nums[ptr]
                    ptr += 1
            nums[ptr], nums[r] = nums[r], nums[ptr]

            if ptr == n - k:
                return nums[ptr]
            elif ptr < n - k:
                return quickselect(ptr + 1, r)
            else:
                return quickselect(l, ptr - 1)

        return quickselect(0, n - 1)

File: leetcode/c/215-kth-largest-element-in-an-array.c
// Author: RT
// Date: 2023-08-15T04:53:33.588Z
// URL: https://leetcode.com/problems/kth-largest-element-in-an-array/description/

int quickselect(int *nums, int n, int k, int l, int r) {
  int ptr = l;
  for (int i = l; i < r; i++) {
    if (nums[i] <= nums[r]) {
      int temp = nums[ptr];
      nums[ptr] = nums[i];
      nums[i] = temp;
      ptr++;
    }
  }
  int temp = nums[ptr];
  nums[ptr] = nums[r];
  nums[r] = temp;

  if (ptr == n - k) {
    return nums[ptr];
  } else if (ptr < n - k) {
    return quickselect(nums, n, k, ptr + 1, r);
  } else {
    return quickselect(nums, n, k, l, ptr - 1);
  }
}

int findKthLargest(int* nums, int numsSize, int k) {
  return quickselect(nums, numsSize, k, 0, numsSize - 1);
}
